src/Analyzers/CSharp/Analyzers/NamespaceFileSync/NamespaceFileSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,177 @@
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.NamespaceFileSync
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class NamespaceFileSyncDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private static readonly LocalizableResourceString s_localizableTitle = new LocalizableResourceString(
+          nameof(CSharpAnalyzersResources.Namespace_named_incorrectly), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        private static readonly LocalizableResourceString s_localizableInsideMessage = new LocalizableResourceString(
+            nameof(CSharpAnalyzersResources.Namespace_must_match_folder_structure), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        public const string RootNamespaceOption = "build_property.RootNamespace";
+        public const string ProjectDirOption = "build_property.ProjectDir";
+
+        public NamespaceFileSyncDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId,
+                   CSharpCodeStyleOptions.PreferNamespaceMatchFolderStructure,
+                   LanguageNames.CSharp,
+                   s_localizableInsideMessage)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNamespaceNode, SyntaxKind.NamespaceDeclaration);
+        }
+
+        private void AnalyzeNamespaceNode(SyntaxNodeAnalysisContext context)
+        {
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(RootNamespaceOption, out var rootNamespace)
+                || rootNamespace is null)
+            {
+                return;
+            }
+
+            if (!context.Options.AnalyzerConfigOptionsProvider.GlobalOptions.TryGetValue(ProjectDirOption, out var projectDir)
+                || projectDir is null or { Length: 0 })
+            {
+                return;
+            }
+
+            if (context.Node is NamespaceDeclarationSyntax namespaceDecl &&
+                IsFileAndNamespaceValid(namespaceDecl, rootNamespace, projectDir, out var targetNamespace) &&
+                IsNamespaceSyncSupported(namespaceDecl, context.SemanticModel))
+            {
+                ReportDiagnostics(context, this.Descriptor, namespaceDecl, ReportDiagnostic.Warn, targetNamespace);
+            }
+        }
+
+        private static bool IsNamespaceSyncSupported(NamespaceDeclarationSyntax namespaceDeclaration, SemanticModel semanticModel)
```

Comment:
in general, order parameters from wider to narrower.  so here, it would be SemanticModel->NamespaceDecl.